File: frontend/src/types/transfer.ts
import type { Account } from './account';
import type { Category } from './category';
import type { Label } from './label';

export interface Transfer {
  id: string;
  fromAmount: number;
  toAmount: number;
  adjustment: number;
  description?: string;
  transactionDate: string;
  fromAccountId: string;
  fromAccount?: Account;
  toAccountId: string;
  toAccount?: Account;
  categoryId?: string;
  category?: Category;
  labelIds?: string[];
  labels?: Label[];
  createdAt: string;
}

export interface CreateTransferRequest {
  fromAmount?: number;
  toAmount?: number;
  adjustment?: number;
  transactionDate: string;
  fromAccountId: string;
  toAccountId: string;
  categoryId?: string;
  labelIds?: string[];
  description?: string;
}

File: src/main/java/com/budget/tracker/model/Transfer.java
package com.budget.tracker.model;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import jakarta.persistence.*;
import lombok.Getter;
import lombok.Setter;

import java.math.BigDecimal;
import java.time.OffsetDateTime;
import java.util.HashSet;
import java.util.Set;
import java.util.UUID;

@Entity
@Table(name = "transfers")
@Getter
@Setter
@JsonIgnoreProperties({ "hibernateLazyInitializer", "handler" })
public class Transfer extends BaseEntity {

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "from_account_id", nullable = false)
    private Account fromAccount;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "to_account_id", nullable = false)
    private Account toAccount;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "category_id")
    private Category category;

    @ManyToMany(fetch = FetchType.LAZY)
    @JoinTable(name = "transfer_labels", joinColumns = @JoinColumn(name = "transfer_id"), inverseJoinColumns = @JoinColumn(name = "label_id"))
    private Set<Label> labels = new HashSet<>();

    @Column(name = "from_amount", nullable = false)
    private BigDecimal fromAmount;

    @Column(name = "adjustment", nullable = false)
    private BigDecimal adjustment = BigDecimal.ZERO;

    @Column(name = "to_amount", nullable = false)
    private BigDecimal toAmount;

    @Column(columnDefinition = "TEXT")
    private String description;

    @Column(name = "transaction_date", nullable = false)
    private OffsetDateTime transactionDate;

    @Column(name = "user_id", nullable = false)
    private UUID userId;
}
